### PR TITLE
drivers: adc: Add support for RaspberryPi Pico ADC

### PR DESCRIPTION
**Pull request by soburi**
Adding RaspberryPi Pico ADC driver.

The ADC has internally connected temperature sensor,
Add property to enable this.

The ADC has a single VREF. VCC usually connects to it,
but it may not be in a case.
Add property to make configurable it.

This driver was created with reference to the adc_emul implementation.

This PR testing with tests/drivers/adc/adc_api.
At this time, twister couldn't with rpi_pico board.

samles/driver/adc is also worked.

### CODE REVIEW

**Comment by yonsch** [comment]
@soburi The STM32 series has a similar temperature sensor, implemented in `drivers/sensor/stm32_temp`. I think it would be best to follow that example for consistency.

**Comment by soburi** [comment]
> @soburi The STM32 series has a similar temperature sensor, implemented in `drivers/sensor/stm32_temp`. I think it would be best to follow that example for consistency.

I understood.
Removing The temperature_sensor option in this PR , And issue another PR to support it as a sensor.

**Comment by yonsch** [comment]
You can keep it in this PR if you want, just do it in a separate commit if you do. The point is to have a compatible and a "sensor" implementation.

**Comment by soburi** [comment]
Hi, @yonsch,

> You can keep it in this PR if you want, just do it in a separate commit if you do. The point is to have a compatible and a "sensor" implementation.

I removed the temperature-related code from this PR.
Because it makes the interface clean. It is just only writing a flag to registers,
I think it is also able to implement in sensor code.

File: dts/bindings/adc/raspberrypi,pico-adc.yaml
Comment (mbolivar-nordic) [inline dts/bindings/adc/raspberrypi,pico-adc.yaml:12]: Why is this here? It looks unnecessary to me.

Comment (soburi) [reply dts/bindings/adc/raspberrypi,pico-adc.yaml]: It's my careless. Removing it.
I implemented without interrupts firstly, and I forgot to remove it.

Comment (mbolivar-nordic) [reply dts/bindings/adc/raspberrypi,pico-adc.yaml]: Thanks!

**Comment by mbolivar-nordic** [comment]
@soburi thank you for your update. Do you think you can also rebase this PR? There is a conflict in rpi_pico.dts.

**Comment by soburi** [comment]
> @soburi thank you for your update. Do you think you can also rebase this PR? There is a conflict in rpi_pico.dts.

Thanks for notifying me. Update done.

**Comment by mbolivar-nordic** [comment]
@nashif @stephanosio it looks like twister may be running into an internal error here:

```
Deleting output directory /__w/zephyr/zephyr/twister-out
Traceback (most recent call last):
  File "./scripts/twister", line 1370, in <module>
    main()
  File "./scripts/twister", line 1107, in main
    tplan.load_from_file(options.load_tests)
  File "/__w/zephyr/zephyr/scripts/pylib/twister/twisterlib.py", line 3466, in load_from_file
    with open(file, "r") as json_test_plan:
FileNotFoundError: [Errno 2] No such file or directory: 'testplan.csv'
```

https://github.com/zephyrproject-rtos/zephyr/runs/6428521792?check_suite_focus=true

Any ideas what may be wrong?

**Comment by soburi** [comment]
@mbolivar-nordic 

> @nashif @stephanosio it looks like twister may be running into an internal error here:
> 
> ```
> Deleting output directory /__w/zephyr/zephyr/twister-out
> Traceback (most recent call last):
>   File "./scripts/twister", line 1370, in <module>
>     main()
>   File "./scripts/twister", line 1107, in main
>     tplan.load_from_file(options.load_tests)
>   File "/__w/zephyr/zephyr/scripts/pylib/twister/twisterlib.py", line 3466, in load_from_file
>     with open(file, "r") as json_test_plan:
> FileNotFoundError: [Errno 2] No such file or directory: 'testplan.csv'
> ```
> 
> https://github.com/zephyrproject-rtos/zephyr/runs/6428521792?check_suite_focus=true
> 
> Any ideas what may be wrong?

I tried to reproduce with my local env.

I found ...

```
  python3 ./scripts/ci/test_plan.py -c origin/${BASE_REF}.. --pull-request
```
seem to generate testplan.json, not testplan.csv.

And run twister wit the testplan.json from command line,
```
 ./scripts/twister --subset 1/10 --load-tests testplan.json ${TWISTER_COMMON} ${PR_OPTIONS}
```

seem to be correctly running.

**Comment by mbolivar-nordic** [comment]
@soburi OK, thank you for your analysis. I think you may need to rebase on top of main after https://github.com/zephyrproject-rtos/zephyr/pull/45642 is merged to get this fix.

**Comment by soburi** [comment]
@mbolivar-nordic 
> @soburi OK, thank you for your analysis. I think you may need to rebase on top of main after #45642 is merged to get this fix.

Rebasing done.

**Comment by mbolivar-nordic** [comment]
That fix is in, but you seem to be running into errors we're facing elsewhere. I think the problem is not with your PR, but with our CI. I'm working on trying to figure this out

**Comment by soburi** [comment]
@mbolivar-nordic 

> That fix is in, but you seem to be running into errors we're facing elsewhere. I think the problem is not with your PR, but with our CI. I'm working on trying to figure this out

OK, I got it.
I'll wait for the to CI stabilize.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (yonsch) [inline dts/arm/rpi_pico/rp2040.dtsi:104]: missing space before the {

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:1]: Comment is not needed

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:24]: I also ran into this when working on ADC long ago. The approach that was taken for the last conflict (the KHZ and MHZ macros) was to patch `hal_rpi_pico` (see https://github.com/zephyrproject-rtos/hal_rpi_pico/commit/191f5ba46fda49523cdaaef27583d1c875ba2c36).
Some of the functions of the ADC API are useless to us (like `adc_init`, which is unaware of pinctrl), but some may be useful. If using these functions would reduce code duplication, I am for using those instead of reimplementing them.

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:27]:
```
#include <hardware/structs/adc.h>
#include <hardware/regs/adc.h>
```

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:54]: Can this be of type `adc_hw_t`? Then you wouldn't need the `ADC_RPI_HW` macro.

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:215]: You have `dev` in this context, you don't need `data->dev`

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:326]: For the other drivers we've named this `idx`

**Review by yonsch — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c:24]: I have issued
https://github.com/zephyrproject-rtos/hal_rpi_pico/pull/2
to avoid conflict between PICO-SDK and Zephyr.

I will update this PR after the changes are approved.

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c]: Fixed it.

File: dts/arm/rpi_pico/rp2040.dtsi
Comment (soburi) [reply dts/arm/rpi_pico/rp2040.dtsi]: Fixed it.

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c]: Replaced with
`#include <hardware/adc.h>`

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c:18]: Replace code by PICO-SDK that can replace with PICO-SDK's function.
If available no function that we expected, I implemented it with a similar interface to PICO-SDK's function.

**Comment by mbolivar-nordic** [comment]
@soburi thank you for your updates. I see that the twister results are OK now, so the CI issues are hopefully solved. Can you please address the compliance problems? I'm copy/pasting them here for your convenience:

```

-:164: ERROR:FUNCTION_WITHOUT_ARGS: Bad function definition - void adc_start_once() should probably be void adc_start_once(void)
#164: FILE: drivers/adc/adc_rpi_pico.c:63:
+static inline void adc_start_once()

-:169: ERROR:FUNCTION_WITHOUT_ARGS: Bad function definition - uint16_t adc_get_result() should probably be uint16_t adc_get_result(void)
#169: FILE: drivers/adc/adc_rpi_pico.c:68:
+static inline uint16_t adc_get_result()

-:174: ERROR:FUNCTION_WITHOUT_ARGS: Bad function definition - bool adc_get_err() should probably be bool adc_get_err(void)
#174: FILE: drivers/adc/adc_rpi_pico.c:73:
+static inline bool adc_get_err()

-:179: ERROR:FUNCTION_WITHOUT_ARGS: Bad function definition - void adc_clear_errors() should probably be void adc_clear_errors(void)
#179: FILE: drivers/adc/adc_rpi_pico.c:78:
+static inline void adc_clear_errors()

-:188: ERROR:FUNCTION_WITHOUT_ARGS: Bad function definition - void adc_enable() should probably be void adc_enable(void)
#188: FILE: drivers/adc/adc_rpi_pico.c:87:
+static inline void adc_enable()

-:440: ERROR:SPACING: space prohibited before that close parenthesis ')'
#440: FILE: drivers/adc/adc_rpi_pico.c:339:
+		IF_ENABLED(CONFIG_ADC_ASYNC, (.read_async = adc_rpi_read_async, )) \
```

**Comment by mbolivar-nordic** [comment]
By the way, unfortunately we did not get this in time for Zephyr v3.1, but hopefully it can be merged very soon after the v3.2 merge window opens.

**Comment by soburi** [comment]
@mbolivar-nordic 
> @soburi thank you for your updates. I see that the twister results are OK now, so the CI issues are hopefully solved. Can you please address the compliance problems? I'm copy/pasting them here for your convenience:

I updated it. Thank you for mentioning the CI error.

> By the way, unfortunately we did not get this in time for Zephyr v3.1, but hopefully it can be merged very soon after the v3.2 merge window opens.

I understand it. I'll rebase if needed after the new version is released.

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
This looks great to me from the devicetree side, with one additional request. Could you please rename the `vref_mv` property to `vref-mv`? In devicetree source files, we usually use `-` instead of `_` in property names. Otherwise, I am ready to approve this from the devicetree side.

Then @anangl and/or @yonsch need to review from the ADC subsystem perspective. Once those approvals are received, we can merge this pull request.

Thanks for your continued effort!

**Comment by soburi** [comment]
Hi, @mbolivar-nordic 
Thanks for comment.
> Could you please rename the vref_mv property to vref-mv?
I fixed it.

**Comment by soburi** [comment]
Add shell support.

**Review by mbolivar-nordic — DISMISSED** [review]
Binding looks OK

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:10]: These will now need the `<zephyr/...>` prefix

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [inline drivers/adc/adc_rpi_pico.c:331]: The comma at the end is better left out of the macro

File: drivers/adc/adc_rpi_pico.c
Comment (yonsch) [reply drivers/adc/adc_rpi_pico.c:333]: I still see `_num`, is it just the diff not working correctly?

**Review by yonsch — CHANGES_REQUESTED** [review]
Looking better, only a few more fixes and then we can merge the PR in the HAL repo

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c:10]: Add <zephyr/~> and reorganized includes.

File: drivers/adc/adc_rpi_pico.c
Comment (soburi) [reply drivers/adc/adc_rpi_pico.c:333]: I forget update it.
It is correct already when now.

Comment (mbolivar-nordic) [reply drivers/adc/adc_rpi_pico.c:333]: > I forget update it.
> It is correct already when now.

I am confused by this response - do you mean you will update it to use `idx`?

Comment (soburi) [reply drivers/adc/adc_rpi_pico.c]: @mbolivar-nordic 

Sorry for the confusion.

I made the fix on June 9.
But the repository was not updated because the push failed.
I uploaded the fix just a few minutes ago.
This comment fix this comment and the following two remarks with this update.

https://github.com/zephyrproject-rtos/zephyr/pull/45363#discussion_r892171981
https://github.com/zephyrproject-rtos/zephyr/pull/45363#discussion_r892177355

**Comment by carlescufi** [comment]
@yonsch @pdgendt please take a look

**Comment by pdgendt** [comment]
@carlescufi, not sure why I got pinged, I have no experience with these changes..

**Review by yonsch — DISMISSED** [review]

**Comment by carlescufi** [comment]
@petejohanson please take a look

**Comment by carlescufi** [comment]
@soburi please rebase

**Comment by soburi** [comment]
Hi, @carlescufi 
> @soburi please rebase

Thanks for the notification.
I done updating. 
tests/drivers/adc/adc_api/src/test_adc.c and samples/drivers/adc/boards/rpi_pico.overlay has some changes to adapt to change adbc83fbf6ae1a89439b726bd1c04c401c89c04b.

**Review by yonsch — DISMISSED** [review]

**Comment by soburi** [comment]
Use DT_HAS_RASPBERRYPI_PICO_ADC_ENABLED instead of $dt_compat_enabled.

**Review by yonsch — DISMISSED** [review]

**Comment by yonsch** [comment]
With SPI merged, this now has conflicts

**Comment by soburi** [comment]
Hi,  @yonsch 
> With SPI merged, this now has conflicts

Thanks for your support!

I'm a bit busy now by another task, so I'll take care of it over the weekend.
Just a moment, please.

**Comment by soburi** [comment]
Hi, @yonsch ,

> With SPI merged, this now has conflicts

Update done.

**Review by yonsch — DISMISSED** [review]

**Comment by soburi** [comment]
Hi, @anangl @mbolivar-nordic @galak @stephanosio @carlescufi @nashif 

I would like to proceed with this PR  (and, also below), so could you give a review this PR if possible?

https://github.com/zephyrproject-rtos/hal_rpi_pico/pull/2#issuecomment-1228116855

As you can see in the conversation log, the points pointed out in the review have solved.
So I think there are no problems.

I would like to put it into the next release if possible.

**Review by nordicjm — COMMENTED** [review]
Looks good, will approve once pull link is replaced with commit ID.

Future update to support the 5th channel (temperature) would be ideal

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
@soburi https://github.com/zephyrproject-rtos/hal_rpi_pico/pull/2 is merged now.

**Review by nordicjm — APPROVED** [review]

**Review by yonsch — APPROVED** [review]

**Comment by yonsch** [comment]
@soburi Are you still planning to work on the temperature sensor?

**Comment by soburi** [comment]
@yonsch 
Oh! I forgot it completely.
Thank you for notification.
I'm a little busy right now, so I think I'll be able to handle it by the end of the year.
If you are already starting to handle it, I will leave it to you.

**Comment by yonsch** [comment]
Thanks @soburi , I didn't start working on it. If you want to finish it yourself go ahead, it's not urgent. If it's too much trouble I can do it no problem.

**Pull request merged**
